### PR TITLE
Parse ast-id

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libyul/AST.h [hunk 1]
```diff
@@ -37,12 +38,22 @@ using Type = YulString;
...
+		langutil::SourceLocation _location = {},
+		std::optional<int64_t> _astID = {}
+	)
 	{
-		return std::make_shared<DebugData const>(_location);
+		return std::make_shared<DebugData const>(std::move(_location), std::move(_astID));
 	}
+
+	langutil::SourceLocation location;
+	/// ID in the source AST.
```
Comment (cameel) [inline libyul/AST.h]: I'd emphasize that it's not the Yul source.
```suggestion
	/// ID in the Solidity source AST.
```

File: libyul/AsmParser.h [hunk 2]
```diff
@@ -114,6 +114,12 @@ class Parser: public langutil::ParserBase
 		langutil::SourceLocation const& _commentLocation
 	);
 
+	std::optional<std::pair<std::string_view, std::optional<int>>>
+	parseASTIDComment(
```
Comment (cameel) [inline libyul/AsmParser.h]: Why not this? It's not that long.
```suggestion
	std::optional<std::pair<std::string_view, std::optional<int>>> parseASTIDComment(
```

File: libyul/AsmParser.cpp [hunk 3]
```diff
@@ -222,6 +231,37 @@ optional<pair<string_view, SourceLocation>> Parser::parseSrcComment(
...
+		R"~~(^(\d+))~~",
+		regex_constants::ECMAScript | regex_constants::optimize
+	);
+	match_results<string_view::const_iterator> match;
+	optional<int> astID;
+	bool matched = regex_search(_arguments.cbegin(), _arguments.cend(), match, argRegex);
+	if (matched)
+	{
+		solAssert(match.size() == 2, "");
+		_arguments = _arguments.substr(static_cast<size_t>(match.position() + match.length()));
```
Comment (cameel) [inline libyul/AsmParser.cpp]:
```suggestion
		tail = _arguments.substr(static_cast<size_t>(match.position() + match.length()));
```

File: test/libyul/Parser.cpp [hunk 4]
```diff
@@ -804,6 +804,67 @@ BOOST_AUTO_TEST_CASE(customSourceLocations_with_code_snippets_with_nested_locati
...
+			function f(x) -> y {}
+			mstore(1, 2)
+		}
+	)";
+	EVMDialectTyped const& dialect = EVMDialectTyped::instance(EVMVersion{});
+	shared_ptr<Block> result = parse(sourceText, dialect, reporter);
+	BOOST_REQUIRE(!!result);
+	BOOST_CHECK(result->debugData->astID == int64_t(7));
+	auto const& funDef = get<FunctionDefinition>(result->statements.at(0));
+	BOOST_CHECK(funDef.debugData->astID == int64_t(2));
```
Comment (cameel) [inline test/libyul/Parser.cpp:824]: Oh, looks like there are no dedicated `int64_t` literals in C++? Apparently the standard way to get such a literal is `INT64_C(7)`. But then I guess that's not much better than a cast.

File: libyul/AsmParser.cpp [hunk 5]
```diff
@@ -222,6 +231,37 @@ optional<pair<string_view, SourceLocation>> Parser::parseSrcComment(
 	return {{tail, SourceLocation{}}};
 }
 
+optional<pair<string_view, optional<int>>> Parser::parseASTIDComment(
+	string_view _arguments,
+	langutil::SourceLocation const& _commentLocation
+)
+{
+	static regex const argRegex = regex(
+		R"~~(^(\d+))~~",
```
Comment (cameel) [inline libyul/AsmParser.cpp]: For consistency with `@src` I think this should require whitespace after a number:
```suggestion
		R"~~(^(\d+)(?:\s+|$))~~",
```

Also, we should have a test for this case.

File: test/libyul/Parser.cpp [hunk 6]
```diff
@@ -804,6 +804,67 @@ BOOST_AUTO_TEST_CASE(customSourceLocations_with_code_snippets_with_nested_locati
...
+	BOOST_CHECK(funDef.parameters.at(0).debugData->astID == nullopt);
+	BOOST_CHECK(debugDataOf(result->statements.at(1))->astID == nullopt);
+}
+
+BOOST_AUTO_TEST_CASE(astid_invalid)
+{
+	ErrorList errorList;
+	ErrorReporter reporter(errorList);
+	auto const sourceText = R"(
+		/// @src -1:-1:-1 @ast-id abc @src 1:1:1
```
Comment (cameel) [inline test/libyul/Parser.cpp:870]: I'd add a test with more than one `@ast-id` in a single comment. The last value still wins in that case, right?

File: libyul/AsmParser.cpp (hunk 5 above)
Comment (cameel) [reply libyul/AsmParser.cpp]: Also, we should have a test case with an ID larger than `int64_t`.

**Review by cameel — COMMENTED** [review]
Looks fine. I see no big problems. Just the whitespace handling is a bit inconsistent with `@src` and I'd add tests for a few more corner cases.

File: libyul/AST.h (hunk 1 above)
Comment (chriseth) [reply libyul/AST.h]: I mean it doesn't have to be solidity...

File: libyul/AsmParser.h [hunk 7]
```diff
@@ -108,8 +108,12 @@ class Parser: public langutil::ParserBase
...
 	void fetchDebugDataFromComment();
 
-	std::optional<std::pair<std::string_view, langutil::SourceLocation>>
-	parseSrcComment(
+	std::optional<std::pair<std::string_view, langutil::SourceLocation>>parseSrcComment(
+		std::string_view _arguments,
+		langutil::SourceLocation const& _commentLocation
+	);
+
+	std::optional<std::pair<std::string_view, std::optional<int>>>parseASTIDComment(
```
Comment (cameel) [inline libyul/AsmParser.h]:
```suggestion
	std::optional<std::pair<std::string_view, langutil::SourceLocation>> parseSrcComment(
		std::string_view _arguments,
		langutil::SourceLocation const& _commentLocation
	);

	std::optional<std::pair<std::string_view, std::optional<int>>> parseASTIDComment(
```

**Review by cameel — DISMISSED** [review]

**Review by cameel — DISMISSED** [review]

File: libyul/AsmParser.h (hunk 7 above)
Comment (chriseth) [reply libyul/AsmParser.h]: argh :)
